src/solvers/dg/dg.jl

```diff
@@ -246,7 +246,25 @@ end
 end
 
 
-function allocate_coefficients(mesh::TreeMesh, equations, dg::DG, cache)
+@inline function wrap_array(u_ode::AbstractVector, mesh::Union{TreeMesh, StructuredMesh}, equations, dg::DG, cache)
+  @boundscheck begin
+    @assert length(u_ode) == nvariables(equations) * nnodes(dg)^ndims(mesh) * nelements(dg, cache)
+  end
+  # We would like to use
+  #   reshape(u_ode, (nvariables(equations), nnodes(dg), nnodes(dg), nelements(dg, cache)))
+  # but that results in
+  #   ERROR: LoadError: cannot resize array with shared data
+  # when we resize! `u_ode` during AMR.
+
+  # The following version is fast and allows us to `resize!(u_ode, ...)`.
+  # OBS! Remember to `GC.@preserve` temporaries such as copies of `u_ode`
+  #      and other stuff that is only used indirectly via `wrap_array` afterwards!
+  unsafe_wrap(Array{eltype(u_ode), ndims(mesh)+2}, pointer(u_ode),
+              (nvariables(equations), fill(nnodes(dg), ndims(mesh))..., nelements(dg, cache)))
```

Comment:
Is there a better way to write `ndims(mesh)` times `nnodes(dg)` than `fill(nnodes(dg), ndims(mesh)...`?

Reply:
Not that I know of. What was the original formulation? I think @ranocha had that already tuned for type stability etc.

Reply:
In the original there were 3 versions, one for each dimension.

Reply:
Instead of constructing an array, you could use an `NTuple`, i.e.,
```julia
ntuple(_ -> nnodes(dg), ndims(mesh))...
```
However, since this is a rather technical core function, I wonder what your rationale is for changing it at all? If it ain't broken, don't fix it...

Reply:
This function was dispatched specifically for the `TreeMesh` and instead of adding `StructuredMesh` to 3 different versions, I rather reduced all three very similar functions to this one.

Reply:
If adding `StructuredMesh` means to change the single type annotation to a `Union`, I'd prefer this change. If you think this is something we should simplify, I'd rather do this in a separate PR (maybe just create an issue?), since this PR is already huge as it is...

Reply:
Alright